### PR TITLE
PhoneIoT Update

### PR DESCRIPTION
**Pull request by dragazo**
This brings in all the PhoneIoT changes I've done while building the iOS app. Most changes are just new optional parameters for custom controls. The only breaking changes are splitting `getState` into `getToggleState` and `isPressed`, as well as renaming an optional parameter for checkboxes from `state` to `checked`.

With this version, I think it's ready for a stable release. The Android app is already on the [play store](https://play.google.com/store/apps/details?id=org.netsblox.phoneiot), and the iOS app is fully functional aside from the "run in background" setting.

### CODE REVIEW

**Comment by brollb** [comment]
Why is the browser submodule changed?

File: src/server/services/procedures/phone-iot/common.js
Comment (brollb) [inline src/server/services/procedures/phone-iot/common.js:88]: These should be added as custom input types rather than custom parsing handled within the service. This will ensure we have standard error messages rather than the ones used here.

File: src/server/services/procedures/phone-iot/phone-iot.js
Comment (brollb) [inline src/server/services/procedures/phone-iot/phone-iot.js:300]: It would be better to make options a parameterized type rather than adding your own parsing code here.

**Comment by dragazo** [comment]
Uhoh, looks like I accidentally committed (the wrong version of) the browser submodule when I merged into master. (curse you, `git add .`). I just added a new commit that brings it back to the same version used by master.

**Review by brollb — CHANGES_REQUESTED** [review]

File: src/server/services/procedures/phone-iot/phone-iot.js
Comment (brollb) [reply src/server/services/procedures/phone-iot/phone-iot.js:300]: (This would probably be good to add in its own PR.)

File: src/server/services/procedures/phone-iot/common.js
Comment (dragazo) [reply src/server/services/procedures/phone-iot/common.js]: New commit moves all of these out to `input-types`, except the options parser, which I might update and generalize as a separate PR since I can change Chart to use it at the same time. I added an input type called `Enum` to implement some of the other types I was using.

**Comment by dragazo** [comment]
@brollb Ok, how does the most recent version look? I addressed the input types issue, unless you meant to do something else. I think the custom options parser would be good as a separate PR since since it would be more time consuming and could be used in some other services.

In the most recent commit, I added some thorough documentation files. In doing so, I realized I wanted one last-minute change before stabilizing the API, so I hid the `getDevices` RPC. It promoted the network attacks like Roboscape used, but for PhoneIoT it seems less reasonable. I figure we can add it back if needed, but once stable we wouldn't have been able to remove it.

**Comment by brollb** [comment]
Nice work! I think this is probably fine for now. I got a little caught up looking at the details for the new enum type but it can probably be addressed in another PR so we don't hold up the deployment of the service.

I still need to flesh my concerns out in more detail but essentially I am wondering:
- How well will this approach fare when used in an RPC annotation (both as a named and anonymous enum)?
- It's not clear if we ever want case-sensitivity (string comparison is case insensitive in NetsBlox and it would be nice to follow this convention)
- Do we want to specify the values for the enum? Why not just a list of valid values? The only value that really seems like it would be affected by this is the new `Bool` type but this might not be a bad thing since just using `true` or `false` would make the error message shorter.
- The error message should probably list off the valid values that are allowed. Perhaps something like `${id} must be one of ${variants.join(', ')}`. I have been trying to make the error messages explicitly share the requirements for a correct value rather than simply that it is incorrect. (For example, `Latitude must be between -90 and 90` as some students didn't know why their latitude of 100 was incorrect since they didn't realize that latitudes had to be in that specific range.)

**Comment by brollb** [comment]
Here are other services/RPCs that could use the enum type:
- Battleship (ship type, direction to face)
- CoreNLP (annotators - list of enums)
- CORGIS (dataset names - this service isn't really supported atm so I wouldn't migrate it)
- COVID19 (country)
- HumanMortalityDatabase (country, gender, category)
- (Probably more :) )

It isn't clear that all these should be should be handled using enums since some have a huge list of valid values (ie, annotators, valid country names).

Also, it's worth noting that this part of the conversation stemmed from seeing custom type parsing logic in a service (a code smell in NetsBlox as it leads to inconsistent error handling across services). However, I am a little concerned about defining service-specific types in the input-types.js file since we would like the logic for a single service to be self-contained. I think the cleanest way to work around this would be to add an API on the input-types which allows services to define their own types.

**Comment by dragazo** [comment]
@brollb Yeah, I think `Enum` could be useful in a lot of places. Here's an unorganized stream of responses to some of your points above. I haven't tried using `Enum` in annotations yet, but I was thinking we would defined a custom type for each one and just use that. I updated `Enum` in the last commit. I removed the case sensitive thing - always insensitive now. I added a little API to define custom types - it has a `defineForService` function for setting your types, and a `forService` function to get them (async because we can't be sure your service loads after theirs). The idea is that you bind the `input-types` module to `types`, so it reads nicely as `types.forService`, etc. I change the error message for `Enum` to list all the valid values (after 10 it gets ellipses). When you mentioned explicit values for the enum, I'm not sure how other services would want it yet, but for PhoneIoT I need this for interop with the mobile app; otherwise I'd just have to do it somewhere else manually. Oh, and I removed the `yes` and `no` values from `Bool`.

**Comment by brollb** [comment]
Cool. I might tweak some things and then merge it. Nice work!

**Comment by dragazo** [comment]
So with the new commit, `Enum` doesn't have mapping behavior (like I had with passing an object)? Would services need to do that separately?

**Comment by brollb** [comment]
Yeah, the problem was making sure it was able to be specified more ergonomically within the RPC annotation. The current approach should be able to specify like this:
```javascript
/**
 * @type {Enum<north,south,east,west>} direction
 */
```

In this case, the `direction` argument is guaranteed to be one of the listed strings. Any mapping would need to be handled by the service :/

**Comment by dragazo** [comment]
It looks like as long as we support arrays, `Enum` can be used with type annotations. I added back the same mapping logic as before and checked that type annotations still worked. So we can conveniently define non-mapping enums in type annotations, and also have the ability to make mapping enums if we call it explicitly.

**Review by brollb — APPROVED** [review]

**Comment by brollb** [comment]
@dragazo - Do you mind updating the options to use the duck typed object syntax so the type checking is handled in a standard way? Then it should be all set!

I am hoping to use this sort of information to generate meaningful documentation for services, too, so it is nice to have the field-level descriptions :) (along with standard error messages).

**Comment by dragazo** [comment]
Sounds good. I just made a commit that adds duck typing for the rest, and removed my `parseOptions` function.

Just in case this comes up elsewhere: I had been using `Object` for taking key-value pairs for `listenToSensors` (but without duck typing). After the merge, this was giving an 'extra fields' error, so I had to explicitly declare all the fields. That's why the most recent commit changed some sensor names (to remove spaces).

The defining custom names is nice, but might lead to name collisions down the road. I imagine there's a context object somewhere when the jsdoc is parsed, so I could look into using the service name to separate them out (different PR, of course).

**Comment by brollb** [comment]
Yeah, I agree about the potential name collisions. That would be nice to clean up but would likely need to be handled in the `defineType` function. (If a type is used across multiple services, it should be placed in the input-types.js file.)

Thanks!

**Pull request merged**
